New dependency resolver does not properly detect a cycle with a transitive dependency with the same name as the root project

### NuGet Product Used

dotnet.exe

### Product Version

9.0.0

### Worked before?

Assume no

### Impact

It's more difficult to complete my work

### Repro Steps & Context

1. Name a project `LightGbm`
2. Try to install https://www.nuget.org/packages/Microsoft.ML.LightGbm

It fails with message 'the given key was not present in the dictionary'

![Image](https://github.com/user-attachments/assets/8b73526a-5379-4197-af68-66e0d5acd5c4)

This is happening because a transitive dependency has the same name as the root project and the new dependency resolver tries to replace the root project with it instead of just ignoring it for later since it needs to be detected as a cycle instead.

### Verbose Logs

```shell

```

Thank you for reporting this issue.  This is caused by a bug in the new dependency resolver.  It should be informing you of a cycle and if you use the legacy dependency resolver you get the proper error message:

```
error NU1108: Cycle detected.
error NU1108:   LightGbm -> Microsoft.ML.LightGbm 4.0.1 -> LightGBM (>= 3.3.5)
```

You would not be able to install a package with a transitive dependency with the same name as your project so you'll need to rename your project.  To workaround the issue, you can also enable the legacy dependency resolver: https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#nuget-dependency-resolver

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItem.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItem.cs
@@ -46,12 +46,12 @@ namespace NuGet.Commands
             /// <summary>
             /// Gets or initializes the <see cref="LibraryDependencyIndex" /> associated with this dependency graph item.
             /// </summary>
-            public LibraryDependencyIndex LibraryDependencyIndex { get; init; } = LibraryDependencyIndex.Invalid;
+            public LibraryDependencyIndex LibraryDependencyIndex { get; init; }
 
             /// <summary>
             /// Gets or initializes the <see cref="LibraryRangeIndex" /> associated with this dependency graph item.
             /// </summary>
-            public LibraryRangeIndex LibraryRangeIndex { get; init; } = LibraryRangeIndex.Invalid;
+            public LibraryRangeIndex LibraryRangeIndex { get; init; }
 
             /// <summary>
             /// Gets or initializes the <see cref="LibraryRangeIndex" /> of this dependency graph item's parent.

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItemIndexer.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItemIndexer.cs
@@ -87,6 +87,16 @@ namespace NuGet.Commands
             private LibraryRangeIndex _nextLibraryRangeIndex = LibraryRangeIndex.Project + 1;
 
             /// <summary>
+            /// Initializes a new instance of the <see cref="DependencyGraphItemIndexer" /> class.
+            /// </summary>
+            /// <param name="libraryDependency">A <see cref="LibraryDependency" /> representing the root project.</param>
+            public DependencyGraphItemIndexer(LibraryDependency libraryDependency)
+            {
+                _libraryDependencyTable.TryAdd(libraryDependency.Name, LibraryDependencyIndex.Project);
+                _libraryRangeTable.TryAdd(libraryDependency.LibraryRange, LibraryRangeIndex.Project);
+            }
+
+            /// <summary>
             /// Indexes a <see cref="LibraryDependency" /> and returns a <see cref="LibraryDependencyIndex" /> associated with the its name.
             /// </summary>
             /// <param name="libraryDependency">The <see cref="LibraryDependency" /> of the library to index.</param>

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItemIndexer.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItemIndexer.cs
@@ -17,11 +17,6 @@ namespace NuGet.Commands
         public enum LibraryDependencyIndex : int
         {
             /// <summary>
-            /// An invalid value used only for default initialization.
-            /// </summary>
-            Invalid = -1,
-
-            /// <summary>
             /// The index of the root project.
             /// </summary>
             Project = 0,
@@ -32,11 +27,6 @@ namespace NuGet.Commands
         /// </summary>
         public enum LibraryRangeIndex : int
         {
-            /// <summary>
-            /// An invalid value used only for default initialization.
-            /// </summary>
-            Invalid = -2,
-
             /// <summary>
             /// The index used as a parent of the root project.
             /// </summary>

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs
@@ -45,7 +45,7 @@ namespace NuGet.Commands
         /// <summary>
         /// A <see cref="DependencyGraphItemIndexer" /> used to index dependency graph items.
         /// </summary>
-        private readonly DependencyGraphItemIndexer _indexingTable = new();
+        private readonly DependencyGraphItemIndexer _indexingTable;
 
         /// <summary>
         /// A <see cref="RestoreCollectorLogger" /> used for logging.
@@ -68,6 +68,11 @@ namespace NuGet.Commands
         private readonly TelemetryActivity _telemetryActivity;
 
         /// <summary>
+        /// Represents the root project as a <see cref="LibraryDependency" />.
+        /// </summary>
+        private readonly LibraryDependency _rootProjectLibraryDependency;
+
+        /// <summary>
         /// Initializes a new instance of the <see cref="DependencyGraphResolver" /> class.
         /// </summary>
         /// <param name="logger">A <see cref="RestoreCollectorLogger" /> to use for logging.</param>
@@ -78,6 +83,16 @@ namespace NuGet.Commands
             _logger = logger;
             _request = restoreRequest;
             _telemetryActivity = telemetryActivity;
+
+            _rootProjectLibraryDependency = new LibraryDependency(
+                new LibraryRange()
+                {
+                    Name = _request.Project.Name,
+                    VersionRange = new VersionRange(_request.Project.Version),
+                    TypeConstraint = LibraryDependencyTarget.Project | LibraryDependencyTarget.ExternalProject
+                });
+
+            _indexingTable = new DependencyGraphItemIndexer(_rootProjectLibraryDependency);
         }
 
         /// <summary>
@@ -133,15 +148,6 @@ namespace NuGet.Commands
             // Stores the list of download results which contribute to the overall success of the graph resolution
             DownloadDependencyResolutionResult[]? downloadDependencyResolutionResults = default;
 
-            // A LibraryDependency representing the root of the graph which is the project itself
-            LibraryDependency mainProjectDependency = new(
-                new LibraryRange()
-                {
-                    Name = _request.Project.Name,
-                    VersionRange = new VersionRange(_request.Project.Version),
-                    TypeConstraint = LibraryDependencyTarget.Project | LibraryDependencyTarget.ExternalProject
-                });
-
             // Create the list of framework/runtime pairs to resolve graphs for.  This method returns the pairs in order with the target framework pairs without runtime identifiers come first
             List<FrameworkRuntimePair> projectFrameworkRuntimePairs = RestoreCommand.CreateFrameworkRuntimePairs(_request.Project, runtimeIds: RequestRuntimeUtility.GetRestoreRuntimes(_request));
 
@@ -182,12 +188,12 @@ namespace NuGet.Commands
                 // Create a DependencyGraphItem representing the root project to add to the queue for processing
                 DependencyGraphItem rootProjectDependencyGraphItem = new()
                 {
-                    LibraryDependency = mainProjectDependency,
+                    LibraryDependency = _rootProjectLibraryDependency,
                     LibraryDependencyIndex = LibraryDependencyIndex.Project,
                     LibraryRangeIndex = LibraryRangeIndex.Project,
                     Suppressions = new HashSet<LibraryDependencyIndex>(),
                     FindLibraryTask = ResolverUtility.FindLibraryCachedAsync(
-                        mainProjectDependency.LibraryRange,
+                        _rootProjectLibraryDependency.LibraryRange,
                         frameworkRuntimePair.Framework,
                         runtimeIdentifier: string.IsNullOrWhiteSpace(frameworkRuntimePair.RuntimeIdentifier) ? null : frameworkRuntimePair.RuntimeIdentifier,
                         context,
@@ -1288,6 +1294,12 @@ namespace NuGet.Commands
                 {
                     LibraryDependency childDependency = chosenResolvedItem.Item.Data.Dependencies[i];
                     LibraryDependencyIndex childLibraryDependencyIndex = chosenResolvedItem.GetDependencyIndexForDependencyAt(i);
+
+                    if (childLibraryDependencyIndex == LibraryDependencyIndex.Project)
+                    {
+                        // Skip any dependency with the same name as the project
+                        continue;
+                    }
 
                     HashSet<LibraryDependency>? runtimeDependencies = default;
 

--- a/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommand_AlgorithmEquivalencyTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommand_AlgorithmEquivalencyTests.cs
@@ -52,6 +52,56 @@ namespace NuGet.Commands.FuncTest
             result.LockFile.Targets[0].Libraries[2].Name.Should().Be("X");
         }
 
+        // LightGbm (project) -> Microsoft.ML.LightGbm -> LightGBM 2.0.0 (package)
+        // This should result in an error with the cycle
+        [Fact]
+        public async Task RestoreCommand_WithCycleContainingProject_FailsAndLogsMessage()
+        {
+            // Arrange
+            using var pathContext = new SimpleTestPathContext();
+            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                pathContext.PackageSource,
+                PackageSaveMode.Defaultv3,
+                new SimpleTestPackageContext("Microsoft.ML.LightGbm", "4.0.1")
+                {
+                    Dependencies =
+                    [
+                        new SimpleTestPackageContext("LightGBM", "3.3.5"),
+                    ]
+                });
+
+            PackageSpec projectSpec = ProjectTestHelpers.GetPackageSpecWithProjectNameAndSpec(
+                "LightGbm",
+                pathContext.SolutionRoot,
+                @"
+                {
+                    ""frameworks"": {
+                    ""net472"": {
+                        ""dependencies"": {
+                                ""Microsoft.ML.LightGbm"": ""4.0.1""
+                        }
+                    }
+                    }
+                }");
+
+            // Act & Assert
+            (var result, _) = await ValidateRestoreAlgorithmEquivalency(pathContext, projectSpec);
+
+            // Additional assert
+            result.Success.Should().BeFalse();
+
+            result.LogMessages.Should().HaveCount(1);
+
+            IAssetsLogMessage logMessage = result.LogMessages.Single();
+            logMessage.Code.Should().Be(NuGetLogCode.NU1108);
+            logMessage.Message.Should().Contain("LightGbm -> Microsoft.ML.LightGbm 4.0.1 -> LightGBM (>= 3.3.5)");
+
+            result.LockFile.Targets.Should().HaveCount(1);
+            result.LockFile.Targets[0].Libraries.Should().HaveCount(1);
+            result.LockFile.Targets[0].Libraries[0].Name.Should().Be("Microsoft.ML.LightGbm");
+            result.LockFile.Targets[0].Libraries[0].Version.Should().Be(new NuGetVersion("4.0.1"));
+        }
+
         // Project 1 -> a 1.0.0 -> b 1.0.0
         //           -> b 2.0.0
         // Does not install b 1.0.0 since it is eclipsed by a direct package reference

